### PR TITLE
Support for Azure CNI Overlay

### LINKED ISSUE #319: Support for `network_plugin_mode`
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description

`network_plugin_mode` was added to v3.35.0 of the azurerm provider. This option allows one to specify `Overlay`, which is currently in preview for the Azure CNI.

### New or Affected Resource(s)/Data Source(s)

azurerm_kubernetes_cluster

### Potential Terraform Configuration

```hcl
From main.tf, the network_profile block, add a variable for network_plugin_mode (with a default of Overlay when the network_plugin is azure):

network_profile {
    network_plugin      = var.network_plugin
    network_plugin_mode = var.network_plugin_mode
    dns_service_ip      = var.net_profile_dns_service_ip
    docker_bridge_cidr  = var.net_profile_docker_bridge_cidr
    load_balancer_sku   = var.load_balancer_sku
    network_policy      = var.network_policy
    outbound_type       = var.net_profile_outbound_type
    pod_cidr            = var.net_profile_pod_cidr
    service_cidr        = var.net_profile_service_cidr

  }
```

### References

_No response_

### PR DESCRIPTION
**Pull request by zioproto**
## Describe your changes

Support for Azure CNI Overlay
https://learn.microsoft.com/en-us/azure/aks/azure-cni-overlay

The product is now GA for Linux nodes only. The terraform preconditions take care of preventing invalid or preview configurations.

## Issue number

Closes #319

## Checklist before requesting a review
- [X] The pr title can be used to describe what this pr did in `CHANGELOG.md` file
- [X] I have executed pre-commit on my machine
- [X] I have passed pr-check on my machine

### CODE REVIEW

**Comment by lonegunmanb** [comment]
Hi @zioproto, thanks for opening this pr, it seems like `network_plugin_mode` feature is also in [preview](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/kubernetes_cluster#network_plugin_mode) right? If so, I'm afraid we need defer this pr to its GA.

**Comment by zioproto** [comment]
The feature is GA:
https://github.com/Azure/AKS/issues/3208
The Terraform docs are not updated.

What is in preview is the upgrade of a cluster from Azure CNI to Azure CNI Overlay

File: main.tf [hunk 1]
```diff
@@ -626,6 +631,15 @@ resource "azurerm_kubernetes_cluster_node_pool" "node_pool" {
       condition     = can(regex("[a-z0-9]{1,8}", each.value.name))
       error_message = "A Node Pools name must consist of alphanumeric characters and have a maximum lenght of 8 characters (4 random chars added)"
     }
+    precondition {
+      condition     = (var.network_plugin_mode == "overlay" && each.value.os_type != "Windows") || (var.network_plugin_mode != "overlay")
+      error_message = "Windows Server 2019 node pools are not supported for Overlay and Windows support is still in preview"
+    }
+
+    precondition {
+      condition     = (var.network_plugin_mode == "overlay" && !regex("^Standard_DC[0-9]+s?_v2$", each.value.vm_size)) || (var.network_plugin_mode != "overlay")
```
Comment (zioproto) [inline main.tf:640]: I made a mistake here.
https://developer.hashicorp.com/terraform/language/functions/regex

I assumed `regex` would return `false` if it did not find a match, but instead `regex` returns an error.

@lonegunmanb do you think I can fix this using `can()` ?
https://developer.hashicorp.com/terraform/language/functions/can

**Comment by zioproto** [comment]
Updating provider docs here:
https://github.com/hashicorp/terraform-provider-azurerm/pull/21917

File: main.tf (hunk 1 above)
Comment (lonegunmanb) [reply main.tf]: Yes, `can()` function is designed for such scenarios.

File: main.tf [hunk 2]
```diff
@@ -451,6 +452,10 @@ resource "azurerm_kubernetes_cluster" "main" {
       condition     = !(var.kms_enabled && var.identity_type != "UserAssigned")
       error_message = "KMS etcd encryption doesn't work with system-assigned managed identity."
     }
+    precondition {
+      condition     = (var.network_plugin_mode == "overlay" && var.network_plugin == "azure") || (var.network_plugin_mode != "overlay")
```
Comment (lonegunmanb) [inline main.tf:456]: Could we simplify the expression?:

```hcl
condition     = var.network_plugin_mode != "overlay" || var.network_plugin == "azure"
```

File: main.tf [hunk 3]
```diff
@@ -626,6 +631,15 @@ resource "azurerm_kubernetes_cluster_node_pool" "node_pool" {
       condition     = can(regex("[a-z0-9]{1,8}", each.value.name))
       error_message = "A Node Pools name must consist of alphanumeric characters and have a maximum lenght of 8 characters (4 random chars added)"
     }
+    precondition {
+      condition     = (var.network_plugin_mode == "overlay" && each.value.os_type != "Windows") || (var.network_plugin_mode != "overlay")
```
Comment (lonegunmanb) [inline main.tf:635]: Could we simplify the expression?:

```hcl
condition     = var.network_plugin_mode != "overlay" || each.value.os_type != "Windows"
```

File: main.tf [hunk 4]
```diff
@@ -626,6 +631,15 @@ resource "azurerm_kubernetes_cluster_node_pool" "node_pool" {
...
+      error_message = "Windows Server 2019 node pools are not supported for Overlay and Windows support is still in preview"
+    }
+
+    precondition {
+      condition     = (var.network_plugin_mode == "overlay" && !can(regex("^Standard_DC[0-9]+s?_v2$", each.value.vm_size))) || (var.network_plugin_mode != "overlay")
```
Comment (lonegunmanb) [inline main.tf:640]: Could we simplify the expression?:

```hcl
condition     = var.network_plugin_mode != "overlay" || !can(regex("^Standard_DC[0-9]+s?_v2$", each.value.vm_size))
```

File: variables.tf [hunk 5]
```diff
@@ -678,6 +678,12 @@ variable "network_plugin" {
   nullable    = false
 }
 
+variable "network_plugin_mode" {
+  type        = string
+  description = "(Optional) Specifies the network plugin mode used for building the Kubernetes network. Possible value is Overlay. Changing this forces a new resource to be created."
```
Comment (lonegunmanb) [inline variables.tf:683]: Could be:

```hcl
description = "(Optional) Specifies the network plugin mode used for building the Kubernetes network. Possible value is `Overlay`. Changing this forces a new resource to be created."
```

**Review by lonegunmanb — CHANGES_REQUESTED** [review]
Thanks @zioproto for the update, a few comments.

**Comment by zioproto** [comment]
I pushed I new commit.

I also corrected `overlay` to `Overlay` in all the preconditions

**Review by lonegunmanb — CHANGES_REQUESTED** [review]
We've got the following error in e2e test:

```text
TestExamplesNamedCluster 2023-05-25T02:01:28Z command.go:185: │ Error: creating/updating "Resource: (ResourceId \"/subscriptions/xxx/resourceGroups/c91104ca6615093b-rg/providers/Microsoft.ContainerService/managedClusters/test-cluster\" / Api Version \"2023-01-02-preview\")": PUT https://management.azure.com/subscriptions/xxx/resourceGroups/c91104ca6615093b-rg/providers/Microsoft.ContainerService/managedClusters/test-cluster
TestExamplesNamedCluster 2023-05-25T02:01:28Z command.go:185: │ --------------------------------------------------------------------------------
TestExamplesNamedCluster 2023-05-25T02:01:28Z command.go:185: │ RESPONSE 409: 409 Conflict
TestExamplesNamedCluster 2023-05-25T02:01:28Z command.go:185: │ ERROR CODE: EtagMismatch
TestExamplesNamedCluster 2023-05-25T02:01:28Z command.go:185: │ --------------------------------------------------------------------------------
TestExamplesNamedCluster 2023-05-25T02:01:28Z command.go:185: │ {
TestExamplesNamedCluster 2023-05-25T02:01:28Z command.go:185: │   "code": "EtagMismatch",
TestExamplesNamedCluster 2023-05-25T02:01:28Z command.go:185: │   "details": [
TestExamplesNamedCluster 2023-05-25T02:01:28Z command.go:185: │     {
TestExamplesNamedCluster 2023-05-25T02:01:28Z command.go:185: │       "code": "Unspecified",
TestExamplesNamedCluster 2023-05-25T02:01:28Z command.go:185: │       "message": "rpc error: code = FailedPrecondition desc = Etag mismatched"
TestExamplesNamedCluster 2023-05-25T02:01:28Z command.go:185: │     }
TestExamplesNamedCluster 2023-05-25T02:01:28Z command.go:185: │   ],
TestExamplesNamedCluster 2023-05-25T02:01:28Z command.go:185: │   "message": "Failed to save managed cluster test-cluster in subscription f7a632a5-49db-4c5e-9828-cd62cb753971, resource group c91104ca6615093b-rg. Operation is not allowed: Another operation is in progress",
TestExamplesNamedCluster 2023-05-25T02:01:28Z command.go:185: │   "subcode": "PutManagedClusterAndAgentPools_FailedPrecondition"
TestExamplesNamedCluster 2023-05-25T02:01:28Z command.go:185: │ }
TestExamplesNamedCluster 2023-05-25T02:01:28Z command.go:185: │ --------------------------------------------------------------------------------
TestExamplesNamedCluster 2023-05-25T02:01:28Z command.go:185: │ 
TestExamplesNamedCluster 2023-05-25T02:01:28Z command.go:185: │ 
TestExamplesNamedCluster 2023-05-25T02:01:28Z command.go:185: │   with module.aks_cluster_name.azapi_update_resource.aks_cluster_post_create,
TestExamplesNamedCluster 2023-05-25T02:01:28Z command.go:185: │   on ../../main.tf line 468, in resource "azapi_update_resource" "aks_cluster_post_create":
TestExamplesNamedCluster 2023-05-25T02:01:28Z command.go:185: │  468: resource "azapi_update_resource" "aks_cluster_post_create" {
TestExamplesNamedCluster 2023-05-25T02:01:28Z command.go:185: │ 
TestExamplesNamedCluster 2023-05-25T02:01:28Z command.go:185: ╵
```

Need further investigation.

**Comment by zioproto** [comment]
@lonegunmanb can we run the e2e tests again please ?

**Comment by lonegunmanb** [comment]
Hi @zioproto the pr failed version upgrade test, would you please merge your branch with the latest `main` branch? Thanks!

**Pull request merged**